src/server/pkg/obj/testing/deployment_test.go

```diff
@@ -0,0 +1,359 @@
+								if requests, ok := resources["requests"]; ok {
+									requests := requests.(map[string]interface{})
+									if _, ok := requests["cpu"]; ok {
+										requests["cpu"] = "0"
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+func (npr *ManifestRewriter) Encode(v interface{}) error {
+	return npr.EncodeTransform(v, nil)
+}
+
+func (npr *ManifestRewriter) EncodeProto(m proto.Message) error {
+	return npr.EncodeProtoTransform(m, nil)
+}
+
+func (npr *ManifestRewriter) EncodeTransform(v interface{}, cb func(map[string]interface{}) error) error {
+	return npr.Encoder.EncodeTransform(v, rewriterCallback(cb))
+}
+
+func (npr *ManifestRewriter) EncodeProtoTransform(m proto.Message, cb func(map[string]interface{}) error) error {
+	return npr.Encoder.EncodeProtoTransform(m, rewriterCallback(cb))
+}
+
+func getPachClient(t *testing.T, kubeClient *kube.Clientset, namespace string) *client.APIClient {
+	// Get the pachd service from kubernetes
+	pachd, err := kubeClient.CoreV1().Services(namespace).Get("pachd", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	var port int32
+	for _, servicePort := range pachd.Spec.Ports {
+		if servicePort.Name == "api-grpc-port" {
+			port = servicePort.NodePort
+		}
+	}
+	require.NotEqual(t, 0, port)
+
+	// Get the IP address of the nodes (any _should_ work for the service port)
+	nodes, err := kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
+	require.NoError(t, err)
+
+	// Minikube 'Hostname' address type didn't work when testing, use InternalIP
+	var address string
+	for _, addr := range nodes.Items[0].Status.Addresses {
+		if addr.Type == "InternalIP" {
+			address = addr.Address
+		}
+	}
+	require.NotEqual(t, "", address)
+
+	// Connect to pachd
+	tu.WaitForPachdReady(t, namespace)
+	client, err := client.NewFromAddress(fmt.Sprintf("%s:%d", address, port))
+	require.NoError(t, err)
+	return client
+}
+
+func makeManifest(t *testing.T, backend assets.Backend, secrets map[string][]byte, opts *assets.AssetOpts) string {
+	manifest := &strings.Builder{}
+	jsonEncoder, err := serde.GetEncoder("json", manifest, serde.WithIndent(2), serde.WithOrigName(true))
+	require.NoError(t, err)
+
+	// Create a wrapper encoder that rewrites NodePort ports to be randomly
+	// assigned so that we don't get collisions across namespaces and can run
+	// these tests in parallel.
+	encoder := &ManifestRewriter{Encoder: jsonEncoder}
+
+	// Use a separate hostpath on the kubernetes host for each deployment
+	hostPath := fmt.Sprintf("/var/pachyderm-%s", opts.Namespace)
+	err = assets.WriteAssets(encoder, opts, backend, assets.LocalBackend, 1, hostPath)
+	require.NoError(t, err)
+
+	err = assets.WriteSecret(encoder, secrets, opts)
+	require.NoError(t, err)
+
+	return manifest.String()
+}
+
+func withManifest(t *testing.T, backend assets.Backend, secrets map[string][]byte, callback func(namespace string, pachClient *client.APIClient)) {
+	namespaceName := tu.UniqueString("deployment-test-")
```

Comment:
Is the idea behind testing in separate namespaces that we can test all of the object storage clients simultaneously? I guess this is an empirical question, but is CI big enough to run like 10(?) Pachyderms simultaneously?

Actually, if so, maybe we should start parallelizing the PPS and auth tests...

Reply:
Since there's a significant amount of waiting, I got quite a bit of benefit from the parallelism locally when testing - a big part of the goal of this was to make these tests parallelizable.  They needed to be in a separate namespace anyway to avoid stepping on a developer's default deploy, so I figured I'd go all out.  On my machine, these tests can all run in under a minute.